src/host/output.cpp

```diff
@@ -499,7 +499,11 @@ void SetActiveScreenBuffer(SCREEN_INFORMATION& screenInfo)
 // TODO: MSFT 9450717 This should join the ProcessList class when CtrlEvents become moved into the server. https://osgvsowi/9450717
 void CloseConsoleProcessState()
 {
+    LockConsole();
+    const auto unlock = wil::scope_exit([] { UnlockConsole(); });

```

Comment:
To quote myself from the PR:
> During testing I found that continuously printing text inside msys2 will cause child processes to only exit slowly one by one every 5 seconds.
> This happens because `CloseConsoleProcessState` calls `HandleCtrlEvent` without holding the console lock. This creates a race condition where most of the time the console IO thread is the one picking up the `CTRL_CLOSE_EVENT`. But that's problematic because the `CTRL_CLOSE_EVENT` leads to a `ConsoleControl` call of type `ConsoleEndTask` which calls back into conhost's IO thread and so you got the IO thread waiting on itself to respond.
> Solution: Don't race conditions.

Reply:
So wait, calling `RundownAndExit` is _safe_ to do while holding the console lock? Isn't it another recipe for a deadlock?

Reply:
I don't think it's safe at all, but the conhost code did exactly that. `Window::ConsoleWindowProc` acquires the console lock and then calls `Window::_CloseWindow` which calls `CloseConsoleProcessState` which calls `ServiceLocator::RundownAndExit`.
In the past without ConPTY `RundownAndExit` was trivial since it only did `ProcessExit` but now with the render flush it has been a disaster recipe for quite a while unfortunately...